Firmware/Src/syscalls.c

```diff
@@ -20,34 +21,40 @@
 #define UART_TX_BUFFER_SIZE 64
 static uint8_t uart_tx_buf[UART_TX_BUFFER_SIZE];
 
-int _write(int file, char *data, int len) {
-  //number of bytes written
-  int written = 0;
-  switch (serial_printf_select) {
-
-    case SERIAL_PRINTF_IS_USB: {
-      // transmit over CDC
-      uint8_t status = CDC_Transmit_FS((uint8_t*)data, len);
-      written = (status == USBD_OK) ? len : 0;
-    } break;
-
-    case SERIAL_PRINTF_IS_UART: {
-      //Check length
-      if (len > UART_TX_BUFFER_SIZE)
-        return 0;
-      // Check if transfer is already ongoing
-      if(huart4.gState != HAL_UART_STATE_READY)
-        return 0;
-      // memcpy data into uart_tx_buf
-      memcpy(uart_tx_buf, data, len);
-      // Start DMA background trasnfer
-      HAL_UART_Transmit_DMA(&huart4, uart_tx_buf, len);
-    } break;
-
-    default: {
-      written = 0;
-    } break;
-  }
-
-  return written;
+int _write(int file, char* data, int len) {
+    //number of bytes written
+    int written = 0;
+    switch (serial_printf_select) {
+        case SERIAL_PRINTF_IS_USB: {
+            // transmit over CDC
+            uint8_t status = CDC_Transmit_FS((uint8_t*)data, len);
+            written = (status == USBD_OK) ? len : 0;
+
+            // Wait for transmission to complete
+            osSemaphoreWait(sem_usb_tx, osWaitForever);
+        } break;
+
+        case SERIAL_PRINTF_IS_UART: {
+            //Check length
+            if (len > UART_TX_BUFFER_SIZE)
+                return 0;
+            // memcpy data into uart_tx_buf
+            memcpy(uart_tx_buf, data, len);
+            // Start DMA background trasnfer
+            HAL_UART_Transmit_DMA(&huart4, uart_tx_buf, len);
+            
+            // Wait for the transmission to complete
+            osSemaphoreWait(sem_uart_dma, osWaitForever);
```

Comment:
Wait on UART TX resource available before transmitting, instead of waiting for transmission to finish.